File: Bloomberg/candy-crush-1d.js
const candyCrush = s => {
    let result = "" // prep for final str value
    let checkAgain = false // whether to check again or not 
    let sArr = s.split("").map(ch => ch.charCodeAt(0)) // str to arr

    // check series
    for(let i = 0; i < sArr.length - 2; i++) {
        let tmp = Math.abs(sArr[i])
        if(tmp === Math.abs(sArr[i + 1]) && tmp === Math.abs(sArr[i + 2])) {
            sArr[i] = sArr[i + 1] = sArr[i +2] = -tmp
            checkAgain = true
        } 
    }

    // update result
    for(let i = 0; i < sArr.length; i++) {
        if(sArr[i] > 0) {
            result += String.fromCharCode(sArr[i])
        }
    }
    
    return checkAgain ? candyCrush(result) : result
}

// console.log(candyCrush("aaabbbc"))
// console.log(candyCrush("ccaabbbaacd"))
// console.log(candyCrush("aaabbbacd"))
// console.log(candyCrush("aabbccddeeedcba"))

// const candyCrush = s => {
//     let result = ""
//     let sArr = s.split("").map(ch => ch.charCodeAt(0))
//     let todo = false

//     for(let i = 0; i < sArr.length - 2; i++) {
//         let tmpValue = Math.abs(sArr[i])
//         if(tmpValue === Math.abs(sArr[i + 1]) && tmpValue === Math.abs(sArr[i + 2])) {
//             sArr[i] = sArr[i + 1] = sArr[i +2] = -tmpValue
//             todo = true
//         } 
//     }

//     for(let i = 0; i < sArr.length; i++) {
//         if(sArr[i] > 0) {
//             result += String.fromCharCode(sArr[i])
//         }
//     }
    
//     return todo ? candyCrush(result) : result
// }

const candyCrush1D = s => {
    let output = ""
    let checkAgain = false
    let strArr = s.split("").map(ch => ch.charCodeAt(0))
    // check same char series
    for(let i = 0; i < strArr.length - 2; i++) {
        let tmpValue = Math.abs(strArr[i])
        if(tmpValue === Math.abs(strArr[i + 1]) && tmpValue === Math.abs(strArr[i + 2])) {
            strArr[i] = strArr[i + 1] = strArr[i + 2] = -tmpValue
            checkAgain = true
        }
    }
    // update output
    for(let el of strArr) {
        if(el > 0) output += String.fromCharCode(el)
    }
    // return or recursion?
    return checkAgain ? candyCrush1D(output) : output
}
// console.log(candyCrush("aaabbbc"))
// console.log(candyCrush("ccaabbbaacd"))
// console.log(candyCrush1D("aaabbbacd"))
// console.log(candyCrush1D("ACCBBBBC"))
// console.log(candyCrush1D("aabbccddeeedcba"))

//// 2 pointer array version
function cca(s) {
    let checkAgain = true
    let sArr = s.split("")
    
    while(checkAgain) {
        checkAgain = false
        let counter = 0
        let left = 0
        let right = 0
        while(right < sArr.length) {
            if(sArr[left] === sArr[right]) {
                counter++
                right++
            } else {
                if(counter >= 3) {
                    sArr.splice(left, right - left)
                    right = right - counter
                    checkAgain = true
                }
                left = right
                counter = 0
            }
        }
    }
    return sArr.join("")
}

// console.log(cc("aaabbbc"))
// console.log(cc("ccaabbbaacd"))
// console.log(cc("aaabbbacd"))
// console.log(cc("ACCBBBBC"))
// console.log(cc("aabbccddeeedcba"))

//// 2 pointer string version
function ccs(s) {
    // let checkAgain = false
    let checkAgain = true    
    while(checkAgain) {
        checkAgain = false
        let counter = 0
        let left = 0
        let right = 0
        while(right < s.length) {
            if(s[left] === s[right]) {
                counter++
                right++
            } else {
                if(counter >= 3) {
                    // s = s.replace(Array(counter).fill(s[left]).join(""), "")
                    s = s.replace(s[left].repeat(counter), "")
                    right = right - counter
                    checkAgain = true
                }
                left = right
                counter = 0
            }
        }
    }
    return s
    // return checkAgain ? ccs(s) : s
}

let str = "abcdefghijklmnopqrstuvwxyzzzyyxxwwvvuuttssrrqqppoonnmmllkkjjiihhggffeeddccbbaaaaa?!?!"

// console.time();
// for(let i = 0; i < 111111; i++) candyCrush1D(str)
// console.log("abs version result :", candyCrush1D(str))
// console.timeEnd();

// console.time();
for(let i = 0; i < 111111; i++) cca(str)
console.log("2 pointer array version result :", cca(str))
// console.timeEnd();

// console.time();
// for(let i = 0; i < 111111; i++) ccs(str)
// console.log("2 pointer string version result :", ccs(str))
// console.timeEnd();

File: Bloomberg/candy-crush-1d.py
def candy_crush(s): # ABCCCCCBBAACDDDCCC -> C
    checkAgain = True # set flag true to begin first check
    while checkAgain: # repeat checking till no string update
        checkAgain = False # set default flag not to repeat
        counter = 0 # set same letter counter
        left = 0 # set 2 pointers for letter comparison range
        right = 0 # set 2 pointers for letter comparison range
        while right < len(s): # iterate to the end of string
            if s[left] == s[right]: # when letter comparison is the same
                counter += 1 # increment counter
                right += 1 # expand right end of the range
                if right == len(s) and counter >= 3: # when the same letter sequence is at the end of string
                    s = s.replace(s[left] * counter, "") # remove repeated chrs
                    left = 0 # reset left end
                    right = 0 # reset right end
                    counter = 0 # reset counter
                    checkAgain = True # flag to re-check updated string
            else: # when letter comparison is different
                if counter >= 3: # when different chr is found and counter >= 3
                    s = s.replace(s[left] * counter, "") # remove repeated chrs
                    right = right - counter # sync right end to left end
                    checkAgain = True # flag to re-check updated string
                left = right # reset pointers to start over with next chr 
                counter = 0 # reset counter
    return s

#### test cases
print(candy_crush("ABCCCCCBBAACDDDCCC"))
string = "abcdefghijklmnopqrstuvwxyzzzyyxxwwvvuuttssrrqqppoonnmmllkkjjiihhggffeeddccbbaaaaa?!?!"
print("result : " + candy_crush(string)) # a?!?!
string = "abcdeeeddccccbbaaaaa!"
print("result : " + candy_crush(string)) # abcbb!
string = "abcdeeeddccccbbaaaaa"
print("result : " + candy_crush(string)) # abcbb
string = "axbcdeeeccccddccbbaaaaa"
print("result : " + candy_crush(string)) # ax
string = "abcccbaaaa"
print("result : " + candy_crush(string)) # abb
string = "abcdefghijklmnopqrstuvwxyzzzyyxxwwvvuuttssrrqqppoonnmmllkkjjiihhggffeeddccbbaa"
print("result : " + candy_crush(string)) # empty string

#### speed test
# import datetime
# t1 = datetime.datetime.now()
# string = "abcdefghijklmnopqrstuvwxyzzzyyxxwwvvuuttssrrqqppoonnmmllkkjjiihhggffeeddccbbaa?!?!"
# for i in range(111111): candy_crush(str)
# print("result : " + candy_crush(string)) # ?!?!
# t2 = datetime.datetime.now()
# print(t2 - t1)



#### window range version

string = "abccccbba"
# [a,b,c,c,c,c,b,b,a] => aa
def candy_crush_1d(s):
    string = list(s) # convert string to list
    for i in range(len(string)): # convert chr to int
        string[i] = ord(string[i])
    checkAgain = False # set flag
    counter = 0 # set counter
    for x in range(len(string) - 2):
        tmpValue = abs(string[x]) # set current abs value to compare
        if tmpValue == abs(string[x+1]) and tmpValue == abs(string[x+2]): # when sequence found
            string[x] = -tmpValue # convert values to negative
            string[x+1] = -tmpValue # convert values to negative
            string[x+2] = -tmpValue # convert values to negative
            checkAgain = True # change flag for re-check
    result = "" # set result string
    for x in string: # convert positive int to chr
        if x > 0:
            result += chr(x)
    return candy_crush_1d(result) if checkAgain else result # check whether to repeat

#### test cases
# print(candy_crush_1d(string)) # aa
string = "abcdefghijklmnopqrstuvwxyzzzyyxxwwvvuuttssrrqqppoonnmmllkkjjiihhggffeeddccbbaa"
print(f"Result of candy_crush => " + candy_crush(string)) # ""
print(f"Result of candy_crush_1d => " + candy_crush_1d(string)) # ""
string = "abcdefghijklmnopqrstuvwxyzzzyyxxwwvvuuttssrrqqppoonnmmllkkjjiihhggffeeddccbbaaa!"
print(f"Result of candy_crush => " + candy_crush(string)) # "a!"
print(f"Result of candy_crush_1d => " + candy_crush_1d(string)) # "a!"
string = "abcdefghijklmnopqrstuvwxyzzzyyxxwwvvuuttssrrqqppoonnmmllkkjjiihhggffeeddccbbaaa"
print(f"Result of candy_crush => " + candy_crush(string)) # "a"
print(f"Result of candy_crush_1d => " + candy_crush_1d(string)) # "a"

import datetime
string = "abcdefghijklmnopqrstuvwxyzzzyyxxwwvvuuttssrrqqppoonnmmllkkjjiihhggffeeddccbbaaa"

t1 = datetime.datetime.now()
for i in range(11111): candy_crush(string)
print("2 Pointer version result : " + candy_crush(string)) # a
t2 = datetime.datetime.now()
print(t2 - t1)

t1 = datetime.datetime.now()
for i in range(11111): candy_crush_1d(string)
print("Window Range version result : " + candy_crush_1d(string)) # a
t2 = datetime.datetime.now()
print(t2 - t1)
